### DISCUSSION TITLE
Store FIDO/Webauthn-Credentials on external FIDO-Server / FIDO-Storage

### DISCUSSION THREAD

**Discussion by denniskniep** [keycloak/keycloak#23101] [Ideas]
# Idea
It would be good to have the option to store FIDO-Credentials on an external FIDO-Server.
The FIDO-Server would be responsible for storing and verifying the FIDO-Credentials. 
This solution should also tackle "storage only" solutions like storing FIDO-Credentials in LDAP  #13817

# Proposal
I would propose to keep the already existing authenticators: [WebAuthnPasswordlessAuthenticator](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticator.java) and [WebAuthnAuthenticator](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java). And modify the implementation in that way that the operations like register, verify etc. are delegated to a new SPI, which needs to be developed (i.e. WebAuthnProvider). There is no standardized REST API protocol for RP to FIDO-Server communication. Therefore the SPI Interface must be as generic as possible. The default implementation is based on the existing internal Keycloak store and verification logic. Having that SPI established, everyone could easily create a plugin and implement the specific FIDO-Server / FIDO-Storage that he wants to use. The benefit is that the whole UI and baseline logic must not be reimplemented.

Example WebAuthnProvider SPI interface:
```
WebAuthnRegistrationRequest createWebAuthnRegistrationRequest(RealmModel realm, UserModel user);

void registerWebAuthnCredential(RealmModel realm, UserModel user, WebAuthnRegistrationResponse registrationResponse);



Collection<WebAuthnCredentials> getWebAuthnCredentials(RealmModel realm, UserModel user);

void updateWebAuthnCredential(RealmModel realm, UserModel user, WebAuthnCredential credential);

void deleteWebAuthnCredential(RealmModel realm, UserModel user, WebAuthnCredential credential);



WebAuthnAuthenticationRequest createWebAuthnAuthenticationRequest(RealmModel realm, UserModel user);

bool isValidWebAuthnAuthenticationResponse(RealmModel realm, UserModel user, WebAuthnAuthenticationResponse response) 
```

# Motivation
Enterprise organisations often rely on external FIDO-Servers or external FIDO-Storage. Because the credentials are stored on specifically secured servers.
Furthermore it is possible, that many applications are using the same registered credentials. This is possible once the applications are hosted below the same domain (i.e. app1.company.org, app2.company.org etc.)
Additionally there might be the demand to reuse the registered credentials for [transaction confirmation](https://fidoalliance.org/white-paper-fido-transaction-confirmation/) by other applications.

# Example
FIDO-Server: [Strongkey](https://github.com/StrongKey/fido2)
FIDO-Storage: LDAP (#13817)

**Comment by mposolda** [comment]
I am not 100% sure if we need this considering that there is not standard for this. It seems that everyone using this will need to implement SPI to implement storage by himself. But that is something, which can be already done? We have user-storage SPI and AFAIK it should be possible to implement storing of some data (like WebAuthn credentials) to that external storage with the usage of userStorage SPI? See for instance `LDAPStorageProvider` from Keycloak codebase or our user-storage provider quickstarts [1] and documentation [2] for more details.

[1] https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension
[2] https://www.keycloak.org/docs/latest/server_development/index.html#_user-storage-spi

**Comment by mposolda** [comment]
@ynojima @mabartos @tnorimat Any feedback from you regarding this?

**Reply by ynojima** [reply]
As @mposolda said, Keycloak already have userStorage SPI, which enables externalizing FIDO-storage.
But this proposal delegates whole WebAuthn registration and authentication process including validating attestation and assertion to external FIDO-server.
This SPI proposal leaves more rooms to the SPI implementations, and this makes a difference between what two SPI can do, but I'm not so sure who really needs this feature.
Speaking of WebAuthn credential sharing between two sub-domains and credential reuse for transaction confirmation, these can be realized with userStorage SPI.

**Reply by denniskniep** [reply]
@ynojima exactly. We want to leverage the APIs for registration, validation etc. of an existing FIDO-Server. It would be awesome if keycloak allow this flexibility in the backend. So that we do not need to copy the entire Webauthn UI and build an dedicated plugin, which would also not benefit from upcoming keycloak internal webauthn improvements.
There are several FIDO-Servers on the market (e.g. https://fidoalliance.org/product-category/server/). 
Using a dedicated FIDO-Server makes sense in specific scenarios e.g. when you are operating multiple IDPs and Apps which support webAuthN and want to rely on a single central implementation. Furthermore if you need to fulfill certain requirements regarding security, regulation and compliance.

For sure it would be fine for me to takeover that implementation task. But first I wanted to check with you if such a PR would be reviewed and accepted.

**Reply by tnorimat** [reply]
I also agree the suggestion by @ynojima and @mposolda . UserStorage SPI could store and verify credentials by external FIDO-storage. It might need to introduce a new SPI for externalizing FIDO-Server. IMO, implementing UserStorage SPI for storing and verifing credentials by external FIDO-storage might be a good first step.

**Reply by dasniko** [reply]
Hi, I just came across this discussion, as I have a similar requirement with one of my customers.

The way through the UserStorageSPI came also to my mind, but AFAIK this means that also the user must have its origin through this SPI. Am I right?

In case I'm using e.g the LDAP implementation, I could easily extend the SPI implementation with my custom "advanced SPI implementation" and just overwrite the methods from the interfaces `CredentialInputValidator` and `CredentialInputUpdater` to be able to use another storage location for WebAuthn credentials. So far, so good.

But what do I have to do in case my users origins are the Keycloak DB itself (e.g. they came across an external IdP) and not through a UserStorage? I tried to use an  UserStorageSPI with only the `CredentialInput*` interfaces and to get credentials for a regular internal user from the SPI, but I failed. Is there a possibility/way to achieve this? Did I do something wrong? Or is this just not possible (at the moment)?

Please correct me, if my assumptions are wrong somewhere in the process... I appreciate any hint into the proper direction. 🙏

**Reply by tnorimat** [reply]
@dasniko Hello. Is the following improvement related to your issue?
https://www.keycloak.org/docs/latest/release_notes/index.html#webauthn-improvements

**Reply by dasniko** [reply]
@tnorimat Thanks for your answer, but no, that's just a feature of the WebAuthn implementation.
I'm talking about the same as the OP - a possibility to store a credential value not in Keycloak, but in a separate/foreign system, e.g. a dedicated FIDO2 server.

I can use the UserStorageSPI for that, yes, but IMHO only for users managed through this UserStorageSPI. If I have internal-only users, I can't use an UserStorageSPI to manage the credential storage.

Basically, I want to use the built-in `webauthn` and `webauthn-passwordless` credentials and authenticators, but having an option to store the credentials secret data in an external dedicated FIDO2 server (like e.g. StrongKey). No matter where the users origin comes from (internal, UserStorageSPI, ...)

**Reply by thomasdarimont** [reply]
I think storage of external Credentials in combination with local / federated users could be done already with a custom `org.keycloak.credential.UserCredentialStore`. In case of a locally stored User one could inherit from `org.keycloak.models.jpa.JpaUserCredentialStore` and add the necessary logic to handle the custom credential. Handling custom credentials for (Storage)Federated users is a bit more work but IMHO also possible by subclassing `org.keycloak.storage.jpa.JpaUserFederatedStorageProvider`.

//edit close but no cigar...
It turns out, that `JpaUserProvider` currently uses the [hardcoded `JpaUserCredentialStore` implementation]( https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L100). If we would change that to a provider lookup, we should be able to support the requirements mentioned above without introducing another SPI. For `org.keycloak.storage.jpa.JpaUserFederatedStorageProvider` we could do the same and reuse the `JpaUserCredentialStore` by default or create a custom `FederatedJpaUserCredentialStore` implementation.

**Reply by dasniko** [reply]
Thanks @thomasdarimont This sounds great and seems to be that what I‘m looking for. 
I‘d also vote for changing the hardcoded instance to a provider lookup/usage.

**Reply by thomasdarimont** [reply]
I gave it a quick spin: I think the following provides initial support for pluggable custom credential storage for local and federated users with minimal changes: https://github.com/keycloak/keycloak/compare/main...thomasdarimont:keycloak:poc/plugable-cred-store?expand=1

With this developers could simply overwrite the create methods of the respective factories (JpaUserProviderFactory,  JpaUserFederatedStorageProviderFactory) and pass in their own user (UserFederated)UserCredentialStore implementations.

This could be enhanced with a dedicated "credentialStore" SPI to make it easier for developers to provide custom implementations.

**Reply by ahus1** [reply]
Thank you @thomasdarimont for this example. I agree this is a great idea to make the code more extendable for custom use cases. 

One general changes I want to see: 

* Instead of passing in a `JpaUserCredentialStore` to `JpaUserProvider`, I'd like to pass only a `UserCredentialStore`. This also moves some code which is IMHO better handled in the `JpaUserCredentialStore`. I saw you did this already when reworking the federation parts.
* Instead of making some methods more accessible in `JpaUserCredentialStore`, I'd rather suggest those who want to extend this to either replace it with their `UserCredentialStore` of choice, or add their behavior using a delegator pattern. 

Unfortunately one newly added class was missing (`JpaUserFederatedCredentialStore.java`), so I recreated it. Here's my code suggestion:
https://github.com/keycloak/keycloak/compare/main...ahus1:keycloak:discussion-23101-store-credentials-externally?expand=1

Some difference of style (to be discussed): 

* Instead of having two constructors, have only one
* Provide a default implementation which throws an exception (as this is a public API). This would then be mentioned in the release notes, and the default implementation would vanish in a future release.

Let me know what you think about this. Feel free to create a PR for this.

**Reply by thomasdarimont** [reply]
Just created a PR with the suggested changes: https://github.com/keycloak/keycloak/pull/27476

**Reply by QSchlegel** [reply]
Hi there,

I wanted to try this by extending the [user-storage-simple](https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension/user-storage-simple) quickstart example to store WebAuthn credentials in a property file.

If I understood the discussion correctly I would have to override the create method of the `JpaUserFederatedStorageProviderFactory` with a `CustomUserFederatedStorageProviderFactory.java` to pass a `CustomUserFederatedCredentialStore` in a `CustomUserFederatedStorageProvider`.

Then it is unclear to me how I can extend the UserStorageProvider's `updateCredential` method to use the `CustomUserFederatedCredentialStore` or if this happens automatically.

I think it is very interesting to manage WebAuthn credentials outside of Keycloak and thanks to @thomasdarimont and @ahus1
